### DISCUSSION TITLE
Send an early response to an action, then do stuff in the background

### DISCUSSION THREAD

**Discussion by arnaudambro** [remix-run/remix#3938] [Show and tell / tips]
Hello

With Express, I used to sometimes send a quick response to a request, then do something in the background, like when receiving a webhook for instance, or when I have some statistics to cache after a form completion.

I used to do something like that

```js
router.post("/form", async (req, res) => {
  // tell the sender that I received
  res.status(200).send({ ok: true });

  // do heavy computation afterwards
  // save my form
  // compute/update some statistics
});
```

however in the `action`, it seems that the `Response` is sent when we `return` the function, is it ?
So if I return, I can't do anything afterwards, is it ?

Or - I'm thinking out loud - could I use `finally` ?

```js
export const action = async () => {
  try {
    // save my form
    return json({ ok: tru })
  } finally {
    // compute statistics
  }
}
```

would that be an appropriate pattern ?

**Comment by herecydev** [comment]
You can do it before returning a response? Does it have to be done after?

**Reply by arnaudambro** [reply]
well yes, that's the whole point of my question ! 🙂
- if it's heavy duty stuff, if I do it before returning a response, then the user has to wait for something useless for him
- if it's a webhook that I have to process, I prefer to send a OK 200 directly then process it afterwards

I'm not looking to find another solution, I'm looking to find out if, within an `action`, or even a `loader` maybe, I can send a `Response`, then do something

**Reply by herecydev** [reply]
> if it's heavy duty stuff, if I do it before returning a response, then the user has to wait for something useless for him

Only if you `await` it. Kick off a promise without awaiting and it's still going to happen but won't block the user.

**Comment by berrylover03** [comment]
Sorry but I have a authentic app on my phone And I have to go through that

**Comment by richardscarrott** [comment]
Your example of using finally would work, but just as in your express route, make sure you catch any errors.

You could also do:

```js
somethingAsync().catch(console.error); // Be sure to catch to avoid unhandled rejections from exiting the server
return json({ ok: true });
```

Or if you don't even want the background processing to even begin until you've returned a response you can do:

```js
queueMicrotask(() => somethingAsync().catch(console.error));
return json({ ok: true }); 
```

However I’d only recommend this if it’s ok for your background processing to fail.

If it’s critical for the action to be handled you’re better off pushing it to a persistent queue which has the ability to retry in the event of a failure / outage, e.g. https://github.com/taskforcesh/bullmq
